core/trino-main/src/main/java/io/trino/sql/planner/AdaptivePlanner.java

```diff
@@ -0,0 +1,426 @@
+            return root;
+        }
+
+        // Add the adaptive plan node recursively where initialPlan remain as it is and optimizedPlan as new currentPlan
+        PlanNode adaptivePlan = addAdaptivePlanNode(idAllocator, initialPlan, optimizationResult.plan(), symbolAllocator.getTypes(), optimizationResult.changedPlanNodes());
+        // validate the adaptive plan
+        try (var ignored = scopedSpan(plannerContext.getTracer(), "validate-adaptive-plan")) {
+            planSanityChecker.validateAdaptivePlan(adaptivePlan, session, plannerContext, typeAnalyzer, symbolAllocator.getTypes(), warningCollector);
+        }
+
+        // Fragment the adaptive plan
+        return planFragmenter.createSubPlans(
+                session,
+                new Plan(adaptivePlan, symbolAllocator.getTypes(), StatsAndCosts.empty()),
+                false,
+                warningCollector,
+                fragmentIdAllocator,
+                new PartitioningScheme(Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()), adaptivePlan.getOutputSymbols()),
+                subPlans,
+                // We do not change the fragment ids which have no changes and are not downstream of the
+                // changed plan nodes. This optimization is done to avoid unnecessary stage restart due to speculative
+                // execution.
+                getUnchangedFragments(adaptivePlan, optimizationResult.changedPlanNodes(), fragmentIdentifierToId));
+    }
+
+    private AdaptivePlanOptimizer.Result optimizePlan(
+            PlanNode plan,
+            SymbolAllocator symbolAllocator,
+            RuntimeInfoProvider runtimeInfoProvider,
+            PlanNodeIdAllocator idAllocator)
+    {
+        AdaptivePlanOptimizer.Result result = new AdaptivePlanOptimizer.Result(plan, Set.of());
+        ImmutableSet.Builder<PlanNodeId> changedPlanNodes = ImmutableSet.builder();
+        for (AdaptivePlanOptimizer optimizer : planOptimizers) {
+            result = optimizer.optimizeAndMarkPlanChanges(
+                    result.plan(),
+                    new PlanOptimizer.Context(
+                            session,
+                            symbolAllocator.getTypes(),
+                            symbolAllocator,
+                            idAllocator,
+                            warningCollector,
+                            planOptimizersStatsCollector,
+                            tableStatsProvider,
+                            runtimeInfoProvider));
+            changedPlanNodes.addAll(result.changedPlanNodes());
+        }
+        return new AdaptivePlanOptimizer.Result(result.plan(), changedPlanNodes.build());
+    }
+
+    private PlanNode addAdaptivePlanNode(
+            PlanNodeIdAllocator idAllocator,
+            PlanNode initialPlan,
+            PlanNode optimizedPlanNode,
+            TypeProvider types,
+            Set<PlanNodeId> changedPlanNodes)
+    {
+        if (changedPlanNodes.contains(optimizedPlanNode.getId())) {
```

Comment:
why are we testing changedPlanNodes for `optimizedPlanNode.getId`. Should be `initialPlan.getId()`, right?